src/snowflake/snowpark/modin/plugin/_internal/align_utils.py

```diff
@@ -93,3 +97,120 @@ def align_axis_0_right(
+            'right': Output frame contains columns from right frame.
+        sort: Sort column axis if True.
+
+    Returns:
+        New InternalFrame after taking union of given frames.
+    """
+    columns1 = frame1.data_columns_index
+    columns2 = frame2.data_columns_index
+
+    if join == "inner":
+        # Preserves the order from calling index.
+        # For example:
+        # pd.Index([3, 1, 2]).intersection(pd.Index([1, 2, 3]) will result in
+        # pd.Index([3, 1, 2])
+        data_column_labels = columns1.intersection(columns2, sort=False)
+    elif join == "outer":
+        # Preserves the order from calling index. And for labels not in calling index
+        # preserves the order from argument index.
+        # For example:
+        # pd.Index([3, 1, 2]).union(pd.Index([1, 4, 2, 3, 5]) will result in
+        # pd.Index([3, 1, 2, 4, 5])
+        data_column_labels = columns1.union(columns2, sort=False)
+    elif join == "left":
+        data_column_labels = columns1
+    elif join == "right":
+        data_column_labels = columns2
+    else:
+        raise AssertionError(
+            f"Invalid join type '{join}'. Accepted values are 'inner', 'outer', 'left', and 'right'."
+        )
+    if sort:
+        data_column_labels = data_column_labels.sort_values()
+
+    frame1 = _select_columns(frame1, data_column_labels.tolist())
+    frame2 = _select_columns(frame2, data_column_labels.tolist())
+    return frame1, frame2
+
+
+def _select_columns(
+    frame: InternalFrame, data_column_labels: list[Hashable]
+) -> InternalFrame:
+    """
+    Select only the given labels from given frame for align. If any data column label is missing
+    in frame add new column with NULL values.
+
+    Args:
+        frame: An InternalFrame
+        data_column_labels: A list of pandas labels.
+
+    Returns:
+        New InternalFrame after only with given data columns.
+
+    """
+    select_list: list[ColumnOrName] = []
+
+    # Add index columns
+    select_list.extend(frame.index_column_snowflake_quoted_identifiers)
+
+    # Add ordering columns
+    select_list.extend(frame.ordering_column_snowflake_quoted_identifiers)
+
+    snowflake_ids = frame.get_snowflake_quoted_identifiers_group_by_pandas_labels(
+        data_column_labels, include_index=False
+    )
+    # Add data columns
+    data_column_snowflake_identifiers = []
+    # A map to keep track number of times a label is already seen.
+    # Native pandas fails with IndexError when either frame has duplicate labels, with
+    # the exception when both frames have exact same lables and exact same order.
+    # In Snowpark pandas, we don't fail concat when duplicates lables are present but
+    # try to match as many columns as possible from the frames.
+    label_count_map: dict[Hashable, int] = {}
+    for label, id_tuple in zip(data_column_labels, snowflake_ids):
+        if len(id_tuple) <= label_count_map.get(label, 0):
```

Comment:
I'm a little confused by what this check is doing. My understanding is:
- `id_tuple` is all the SF quoted identifiers corresponding to the pandas label
- `label_count_map` maps pandas labels to the number of times we've encountered it within a frame
Is `len(id_tuple) <= label_count_map.get(label, 0)` just checking whether this is the first time we've encountered this label in the loop? Why are we choosing to add a NaN column when that's the case, and adding the original column of any duplicates we encounter?

Reply:
`len(id_tuple) <= label_count_map.get(label, 0)` is true when a label does not exist in the original frame (for example when frame has columns A, B, B, D and data_column_labels=[A, B, C, D]. The snowflake_ids is [("A",), ("B", "B_7wi3"), ("B", "B_7wi3"), (), ("D",)]

So the id_tuple for C would be () with len 0 which is <=  label_count_map.get(C, 0). In this case, we would add a NaN column for label C.